beacon_node/beacon_chain/src/otb_verification_service.rs

```diff
@@ -0,0 +1,348 @@
+                .map_err(Error::StoreError)?;
+        }
+    }
+
+    // ensure finalized canonical otb are valid, otherwise kill client
+    for otb in finalized_canonical_otbs {
+        match chain.get_block(otb.root()).await {
+            Ok(Some(block)) => {
+                match validate_merge_block(chain, block.message()).await {
+                    Ok(()) => {
+                        // merge transition block is valid, remove it from OTB
+                        otb.remove_from_store::<T, _>(&chain.store)
+                            .map_err(Error::StoreError)?;
+                        info!(
+                            chain.log,
+                            "Validated merge transition block";
+                            "block_root" => ?otb.root(),
+                            "type" => "finalized"
+                        );
+                    }
+                    Err(BlockError::ExecutionPayloadError(
+                        ExecutionPayloadError::InvalidTerminalPoWBlock { .. },
+                    )) => {
+                        // Finalized Merge Transition Block is Invalid! Kill the Client!
+                        crit!(
+                            chain.log,
+                            "Finalized merge transition block is invalid!";
+                            "msg" => "You must use the `--purge-db` flag to clear the database and restart sync. \
+                            You may be on a hostile network.",
+                            "block_hash" => ?block.canonical_root()
+                        );
+                        let mut shutdown_sender = chain.shutdown_sender();
+                        if let Err(e) = shutdown_sender.try_send(ShutdownReason::Failure(
+                            INVALID_FINALIZED_MERGE_TRANSITION_BLOCK_SHUTDOWN_REASON,
+                        )) {
+                            crit!(
+                                chain.log,
+                                "Failed to shut down client";
+                                "error" => ?e,
+                                "shutdown_reason" => INVALID_FINALIZED_MERGE_TRANSITION_BLOCK_SHUTDOWN_REASON
+                            );
+                        }
+                    }
+                    _ => {}
+                }
+            }
+            Ok(None) => return Err(Error::NoBlockFound(otb)),
+            // Our database has pruned the payload and the payload was unavailable on the EL since
+            // the EL is still syncing or the payload is non-canonical.
+            Err(BeaconChainError::BlockHashMissingFromExecutionLayer(_)) => (),
+            Err(e) => return Err(Error::BeaconChain(e)),
+        }
+    }
+
+    // attempt to validate any non-finalized canonical otb blocks
+    for otb in unfinalized_canonical_otbs {
+        match chain.get_block(otb.root()).await {
+            Ok(Some(block)) => {
+                match validate_merge_block(chain, block.message()).await {
```

Comment:
`validate_merge_block` will return `Ok(())` when it's past `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` and the EL is still syncing. In this scenario we're declare that the OTB is fully valid, when we actually just optimistically verified it again. It would be an OOTB!

I've suggest a fix and regression test for this in https://github.com/paulhauner/lighthouse/commit/a030175a4c854ddeb77c828d16c45ba882dd02a7.

Reply:
Great catch! 🙏
I knew this would be called recursively, which was [why I added this check](https://github.com/sigp/lighthouse/blob/3d5660350171c3dd5b728782c765f8761d3338d0/beacon_node/beacon_chain/src/otb_verification_service.rs#L49-L51) but somehow just didn't realize that it returns `Ok(())` after doing that 🤦‍♂️

I've merged your fix in. I wonder if we should remove that check? I'm not even sure if leveldb returns an error if the entry already exists but that actually would've been better in this case.

Reply:
> I wonder if we should remove the double insert check? I'm not even sure if leveldb returns an error if the entry already exists but that actually would've been better in this case.

I don't think the check is strictly necessary, but I don't mind leaving it there. I don't think it'll have any functional impact.